plugins/thunderbolt-power/fu-plugin-thunderbolt-power.c

```diff
@@ -32,8 +37,104 @@ struct FuPluginData {
 	gboolean       needs_forcepower;
 	gboolean       updating;
 	guint          timeout_id;
+	gboolean       bolt_supported;
+	gint           bolt_fd;
 };
 
+static void
+fu_plugin_thunderbolt_power_bolt_supported (FuPlugin *plugin)
+{
+	FuPluginData *data = fu_plugin_get_data (plugin);
+	g_autoptr(GDBusConnection) connection = NULL;
+	g_autoptr(GVariant) val = NULL;
+	g_autoptr(GDBusProxy) proxy = NULL;
+	g_autoptr(GError) error_local = NULL;
+
+	connection = g_bus_get_sync (G_BUS_TYPE_SYSTEM, NULL, &error_local);
+	if (connection == NULL) {
+		g_warning ("Failed to initialize d-bus connection: %s",
+			   error_local->message);
+		return;
+	}
+
+	proxy = g_dbus_proxy_new_sync (connection,
+					G_DBUS_PROXY_FLAGS_NONE,
+					NULL,
+					BOLT_DBUS_SERVICE,
+					BOLT_DBUS_PATH,
+					BOLT_DBUS_INTERFACE,
+					NULL,
+					&error_local);
+	if (proxy == NULL) {
+		g_warning ("Failed to initialize d-bus proxy: %s",
+			   error_local->message);
+		return;
+	}
+	val = g_dbus_proxy_call_sync (proxy,
+					"ListGuards",
+					NULL,
+					G_DBUS_CALL_FLAGS_NONE,
+					-1,
+					NULL,
+					NULL);
+	data->bolt_supported = (val != NULL);
```

Comment:
Currently that call will return successfully with an empty array even if `boltd` did not detect support for force power, i.e. the wmi sysfs device is missing. I could change it to return an error (NOT_SUPPORTED or something) in that case, but I wonder if I should not change the "PowerState" enum property to include a `unsupported` value. Any opinions?

Reply:
I was wondering if it should be a different plugin entirely...

Reply:
Hmm, keeping them together in the same fwupd plugin is actually ideal in my mind.  The whole reason for this bolt feature is so that only one userspace is accessing the kernel method to avoid software stomping on each other.  If bolt is there, it should be used.  If not, use the kernel directly.  Eventually when new enough bolt is everywhere that we care about, I would actually prefer to drop the direct kernel support from fwupd.

If bolt goes into it's own plugin (say `bolt-thunderbolt-power` and this plugin renames to `kernel-thunderbolt-power`) the model for when to use the `kernel-thunderbolt-power` plugin becomes more complicated.  There isn't really a device registered signal to look for, so how will fwupd know which to prefer?

Reply:
I was thinking about this some more this morning, and I think you're right - we need some assurance from bolt that it can actually force power.  Returning an `unsupported` from `PowerState` property is sufficient as long as a fresh `GDbusProxy` is used every time it's checked.

Reply:
@superm1 I will go and implement that now then. I also think that `GDBusProxy` should keep track of property updates, but in any case a manual call to the  `org.freedesktop.DBus.Properties.Get` method should always work, i.e. something like:

```
g_dbus_proxy_call (proxy, 
                   "org.freedesktop.DBus.Properties.Get", 
                   g_variant_new ("(ss)", BOLT_DBUS_INTERFACE, "PowerState"),
[...]
```

Reply:
Rather than the client need to know to look for magic PowerState property enums (ie !=`unsupported`), maybe a third method `ForcePowerSupported` would be better?  Then that can instantly check whether the kernel module is loaded and give a boolean answer.

Up to you, i'll adjust the PR for whichever you decide.

Reply:
I thought about this some more and decided to extract the whole force power related properties/methods into its own interface: [`org.freedesktop.bolt1.Power`](https://gitlab.freedesktop.org/bolt/bolt/merge_requests/101/diffs#9bddbd9ad5b5362558d3959f3ab59de2a55ee538_205_212). Both methods (`ForcePower`, `ListGuards`), the property (`PowerState` as `State`) and additionally a new boolean `Supported` property are exposed there (and the methods removed from the `bolt1.Manager` interface). I think for this PR changing `BOLT_DBUS_INTERFACE` to `org.freedesktop.bolt1.Power` should suffice;  the new `Supported` boolean, that can be used to check if force-powering is supported or not, should come in handy. 

I am also add support for monitoring changes to the `intel-wmi-thunderbolt` module so that the `Supported` property is updated accordingly. For this I am using the `bind` and `unbind` uevents (see [here](https://gitlab.freedesktop.org/bolt/bolt/merge_requests/101/diffs?commit_id=051011b04c4402ba508f57ecfa2ff335e8365089)) — I hope that is a good way to do that. 

One more thing: it seems to be totally possible to unload the module, even with `force_power` is set to 1 and `boltd` will not yet do the right thing in that case. This needs a few more changes.